File: 02-two_pointers/00042-Trapping_Rain_Water/00042-trapping_rain_water.py
# 42: Trapping Rain Water
# https://leetcode.com/problems/trapping-rain-water/


class Solution:
    # SOLUTION
    def trap(self, height: list[int]) -> int:
        l, r = 0, len(height) - 1
        maxLeft, maxRight = height[l], height[r]
        result = 0

        while l<r:
            if maxLeft<=maxRight:
                l+=1
                maxLeft = max(maxLeft, height[l])
                result += maxLeft - height[l]
            else:
                r-=1
                maxRight = max(maxRight, height[r])
                result += maxRight - height[r]

        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    height = [4,2,0,3,2,5]
    
    # OUTPUT
    result = o.trap(height)
    print(result)

File: 02-two_pointers/00042-Trapping_Rain_Water/00042-trapping_rain_water.cpp
// 42: Trapping Rain Water
// https://leetcode.com/problems/trapping-rain-water/


#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int trap(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;

        int maxLeft = height[l];
        int maxRight = height[r];

        int result = 0;

        while (l<r) {
            if (maxLeft<=maxRight) {
                l++;
                maxLeft = max(maxLeft, height[l]);
                result += maxLeft - height[l];
            } else {
                r--;
                maxRight = max(maxRight, height[r]);
                result += maxRight - height[r];
            }
        }
        
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> height = {4,2,0,3,2,5};
 
    // OUTPUT
    auto result = o.trap(height);
    cout<<result<<endl;

    return 0;
}